### ISSUE TITLE
Enable back the runtime, optionally?

### ISSUE THREAD

**Issue by p0nce** [AuburnSounds/Dplug#203]
Considering #200 (runtime can be enabled back) :
- [x] ~~Find a way to opt-in for the runtime with a `version`, or dub `subConfiguration`.~~  => no need with "Runtime sections"
- [x] initialize the runtime, attach threads etc. without the old OSX work-around
- [x] ~~Think about what to do for GC objects and anything emplacing (roots must be dealed with). Should the audio thread allowed to hold GC roots?~~ => GC object can only created in "Runtime sections". GC objects that must outlive the section must be explicitely rooted.
- [x] say that LDC 1.3 is the minimum supported LDC on OSX (previously: 1.2). => Only if using those "Runtime sections" (EDIT: 1.8 is the minimum now)
- [x] Stop supporting DMD on OSX, it is still not possible to enable the runtime? (EDIT: still not possible)
- [x] Test on major hosts (no need, this is purely testable)
- [x] Test on OSX < 10.12 (EDIT: wrote a warning message in the example instead)

**Comment by p0nce** [comment]
mixed runtime and @nogc still create problems. For example if you mallocEmplace some object which hold a GC object, the GC object won't be reachable and destroyed at the first cycle. (just made this mistake in dplug:host).
If the runtime comes back there will be a need for strong policies for roots.

**Comment by p0nce** [comment]
The "Runtime sections" solution. 

**tl:dr;** You guarantee that GC objects only get created in such a section and that the important ones are referred manually through roots.

how it works: mallocNew still doesn't make roots. Runtime gets initialized as needed, opt-in.T

The whole plugin is still @nogc, BUT in this section we break @nogc and lazily enable the runtime.

**How it might work**

At the end of the section, the runtime is still enabled, so needed objects don't get collected. But _no thread stay attached_.

Roots would be explicit into a "root pointer" or something.

Having no attached threads means the 150 callbacks in the wild don't get to take a mutex to attach/detach threads.

**Comment by p0nce** [comment]
example:

```d

class C
{
    void normalPluginFunction() nothrow @nogc
    {
        MyClass a = runtimeSection(&functionWhoCanUseTheRuntime);
    }

    void functionWhoCanUseTheRuntime()
    {
        import vibed;
        auto a = httpGet("example.com");
        return makeRoot!MyClass(a); // a must outlive this runtime section, find a way
    }
}


// magic here

auto runtimeSection(T)(T delegate() dg)
{
    RuntimeRAIIStruct s; // tasked to attach thread if not already attached, enable runtime lazily
    s.initialize();
    return dg();
}
```

**Comment by p0nce** [comment]
At first limitation will be stronger, with the runtime being init/finalized on each runtime section, hence nothing on the heap will be allowed at the end of the section.

**Comment by p0nce** [comment]
A problem is that to forward delegate and return a delegate enclosed with a RuntimeSection RAII you need to use a closure.

**Comment by p0nce** [comment]
https://dpaste.dzfl.pl/cf44417c98f9

**Comment by p0nce** [comment]
Runtime cannot be initialized after terminate, so it will be a long-lived entity owned by the `Client` subclass
```d
// <needed for runtime> This is required so that the rest of the plug-in can make runtime calls.
ScopedRuntime _runtime;
this()
{
    _runtime.initialize();
}
// </needed for runtime>
```

**Comment by p0nce** [comment]
New example in `Dplug/tests/using-runtime`

**Comment by p0nce** [comment]
Everything works but Issue #291 hides it

**Comment by p0nce** [comment]
Available in v7.1.0

**Comment by p0nce** [comment]
Use at your own risk.
